### PR TITLE
HLRC: Reorg clients package structure

### PR DESCRIPTION
**Pull request by hub-cap**
The Request and Response classes are being moved out of protocol and in
to the HLRC. This commit gets the package structure ready for the move,
as well as moving all of the clients into the new package structure.

### CODE REVIEW

**Comment by hub-cap** [comment]
The structure looks like the following below now. Note that ml has its own set of request converters already, and a further refactor PR will do that to all of the clients. Also something to note is that index already has a few responses, so you can see the structure of that as well. In the end (follow up PR), the plan is to move all of the protocol request/responses into their appropriate packages here and nuke protocol form orbit. I just wanted to get only the structure reorg up so people can see it and comment.

```
client/rest-high-level/src/main/java/org/elasticsearch/client
├── cluster
│   └── ClusterClient.java
├── graph
│   └── GraphClient.java
├── index
│   ├── GetAliasesResponse.java
│   ├── IndicesClient.java
│   └── SyncedFlushResponse.java
├── ingest
│   └── IngestClient.java
├── license
│   └── LicenseClient.java
├── migration
│   └── MigrationClient.java
├── ml
│   ├── MachineLearningClient.java
│   └── MLRequestConverters.java
├── RequestConverters.java
├── RestHighLevelClient.java
├── snapshot
│   └── SnapshotClient.java
├── tasks
│   └── TasksClient.java
├── Validatable.java
├── ValidationException.java
├── watcher
│   └── WatcherClient.java
└── xpack
    └── XPackClient.java
```

File: client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java [hunk 1]
```diff
@@ -1216,7 +1216,7 @@ static Request getMigrationAssistance(IndexUpgradeInfoRequest indexUpgradeInfoRe
         return request;
     }
 
-    static HttpEntity createEntity(ToXContent toXContent, XContentType xContentType) throws IOException {
+    public static HttpEntity createEntity(ToXContent toXContent, XContentType xContentType) throws IOException {
```
Comment (javanna) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java]: I think that if we make all these methods public they should also have javadocs. I think it's a good move that they become public, users could potentially use them without using the high-level client to fire requests, but they become part of our public API this way and we need to be sure that we want that. Maybe though if you plan to split this into different classes belonging each one to their package, the visibility can go back to what it was after that?

File: client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java [hunk 2]
```diff
@@ -131,14 +131,14 @@
 import java.util.StringJoiner;
 import java.util.stream.Collectors;
 
-final class RequestConverters {
-    static final XContentType REQUEST_BODY_CONTENT_TYPE = XContentType.JSON;
+public final class RequestConverters {
```
Comment (javanna) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java]: we should add javadocs to this class if it becomes public

File: client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java [hunk 3]
```diff
@@ -1561,11 +1561,11 @@ static XContentType enforceSameContentType(IndexRequest indexRequest, @Nullable
     /**
      * Utility class to build request's endpoint given its parts as strings
      */
-    static class EndpointBuilder {
+    public static class EndpointBuilder {
```
Comment (javanna) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java]: maybe if Params and EndpontBuilder are also used externally, they should be extracted as top-level classes? WDYT?

File: client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java [hunk 4]
```diff
@@ -1227,7 +1238,7 @@ private static RequestOptions optionsForHeaders(Header[] headers) {
         return options.build();
     }
 
-    static boolean convertExistsResponse(Response response) {
+    public static boolean convertExistsResponse(Response response) {
```
Comment (javanna) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java]: add javadocs?

File: client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java [hunk 5]
```diff
@@ -965,11 +976,11 @@ public final void fieldCapsAsync(FieldCapabilitiesRequest fieldCapabilitiesReque
      * layer has been added to the ReST client, and requests should extend {@link Validatable} instead of {@link ActionRequest}.
      */
     @Deprecated
-    protected final <Req extends ActionRequest, Resp> Resp performRequestAndParseEntity(Req request,
-                                                                            CheckedFunction<Req, Request, IOException> requestConverter,
-                                                                            RequestOptions options,
-                                                                            CheckedFunction<XContentParser, Resp, IOException> entityParser,
-                                                                            Set<Integer> ignores) throws IOException {
+    public final <Req extends ActionRequest, Resp> Resp performRequestAndParseEntity(Req request,
```
Comment (javanna) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java]: One of the side-effects of these methods being public is that they will be suggested to our users by IDEs although they will never be what users need to call, also they become part of our public API while we considered them internal until now. I wish there was a way to not make them public. Maybe we should consider keeping the clients all on the same package? I think that having the other clients subclass this class is a no-go given that the main client needs to have knowledge of all of the clients.

File: client/rest-high-level/src/main/java/org/elasticsearch/client/cluster/ClusterClient.java [hunk 6]
```diff
@@ -41,7 +44,7 @@
 public final class ClusterClient {
     private final RestHighLevelClient restHighLevelClient;
 
-    ClusterClient(RestHighLevelClient restHighLevelClient) {
+    public ClusterClient(RestHighLevelClient restHighLevelClient) {
```
Comment (javanna) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/cluster/ClusterClient.java]: I worry that we open up the possibility for manually creating the sub-clients, potentially even providing a subclass of our main client. I am paranoid I know, but ideally we'd control this. After all there is no reason to open this up to users, it's another side-effect of the package/visibility changes.

File: client/rest-high-level/src/main/java/org/elasticsearch/client/index/GetAliasesResponse.java [hunk 7]
```diff
@@ -17,7 +17,7 @@
  * under the License.
  */
 
-package org.elasticsearch.client;
+package org.elasticsearch.client.index;
```
Comment (javanna) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/index/GetAliasesResponse.java]: this is a breaking change, as well as the existing clients package changes. Are we sure we want to break existing users? It's only a matter of changing imports but still... if we do want to go ahead, we should document it somewhere.

**Review by javanna — COMMENTED** [review]
I left some comments around methods visibility for users and javadocs.

File: client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java (hunk 1 above)
Comment (hub-cap) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java]: some of these method will move in a PR i plan on doing that splits up the request converters, but things like createEntity is shared by all of the clients, so it probably makes sense to add javadoc.

File: client/rest-high-level/src/main/java/org/elasticsearch/client/index/GetAliasesResponse.java (hunk 7 above)
Comment (hub-cap) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/index/GetAliasesResponse.java]: where do we document things like this? do i add a `>breaking` and thats it? do i need to explain it more to whomever is writing the changelog?

**Comment by hub-cap** [comment]
I think @javanna comments are enough to ask if we even want to move the clients. I think it might make more sense to leave them alone so we do not have to muck with the public methods to confuse implementors. Thoughts cc @javanna @nik9000

File: client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java [hunk 8]
```diff
@@ -131,14 +131,14 @@
...
-    static final XContentType REQUEST_BODY_CONTENT_TYPE = XContentType.JSON;
+public final class RequestConverters {
+    public static final XContentType REQUEST_BODY_CONTENT_TYPE = XContentType.JSON;
 
     private RequestConverters() {
         // Contains only status utility methods
     }
 
-    static Request cancelTasks(CancelTasksRequest cancelTasksRequest) {
+    public static Request cancelTasks(CancelTasksRequest cancelTasksRequest) {
```
Comment (nik9000) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java]: I talked to @hub-cap about these - these aren't going to stay public. We'll un-public them in a few PRs.

File: client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java (hunk 2 above)
Comment (nik9000) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java]: But, yeah, this class should have javadocs. And the actual public methods should shift to the top, but maybe that is a nice, small, mechanical change to do in a follow up PR.

File: client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java (hunk 3 above)
Comment (nik9000) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java]: I think they are really only useful in the context of other request converters so I'm fine with them staying inner classes.

File: client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java (hunk 5 above)
Comment (nik9000) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java]: I'd *really* prefer to keep this protected if possible.

File: client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java (hunk 8 above)
Comment (hub-cap) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java]: yes the thought is to move all of the client specific request converters into their own classes and then put them in the same package, removing the `public`. The only public ones will be the methods shared by more than one of the request converters.

**Comment by nik9000** [comment]
I think it is important to keep the public interface of the HighLevelRestClient clear of implementation details of the client. And yet we want folks to be able to use the client. I wonder if something like this is better:

```
/**
 * Utilities for implementing the high level client and extension to it.
 */
public final class HighLevelRestClientCore {
  public <Req extends ActionRequest, Resp> Resp performRequestAndParseEntity(Req request, ...) {...}
  public <Req extends ActionRequest, Resp> Resp performRequest(Req request, ...) {...}
  ...
}

public class RestHighLevelClient implements Closeable {
  private final HighLevelRestClientCore core;

  public RestHighLevelClient(RestHighLevelClientBuilder builder) {
    core = new HighLevelRestClientCore(builder.build());
    ....
  }

  /**
   * Build an area client.
   */ /// I think this should be here so extension to our client can build their own extensions and we don't have to make `core` protected
  protected final <T> T buildSubClient(Function<HighLevelRestclientCore, T> ctor) {
    return ctor.apply(core);
  }

  public final IndicesClient indices() {
    return buildSubClient(IndicesClient::new); // There isn't any state in these so we can build them as we need them and throw them away
  }
}
```


I know the names aren't great, but if we have something like the `HighLevelRestClientCore` then we can document it and folks can use it to extend the rest client. But it won't be public and won't show up when someone pulls up intelli-sense for the client. I think the extra class is worth it to keep the interface clean.


I'm perfectly fine with making those methods in RequestConverters public so long as we intend to make the methods that aren't useful for extension to the client package private again in a follow up.

**Comment by javanna** [comment]
This last proposal sounds good to me. The only small thing I would try and improve is the fact that those internal methods are all made public. Those are really methods that only client objects  need to use. Would it be an option to have a new base abstract class for all the clients that holds all of those methods? Say `RestHighLevelClient` would extend your `HighLevelRestClientCore` and all the other clients would as well. Those utility methods could stay protected at that point.

**Comment by nik9000** [comment]
> Say RestHighLevelClient would extend your HighLevelRestClientCore and all the other clients would as well.

That is fine with me. At that point it'd be `AbstractRestHighLevelClient` and it'd just wrap the `RestClient`, right?

**Comment by javanna** [comment]
yes @nik9000 I think so

**Comment by hub-cap** [comment]
Since this affects the RequestConverters a bunch, I am going to first split those up in a PR and then follow up with this PR. It will cause less churn in the method scope (and I wont forget to un-public any of those methods again).

**Comment by hub-cap** [comment]
PR is here #33281 for the RequestConverter split up.

**Comment by hub-cap** [comment]
the conflicts are quite large here, so I am going to close this in favor of a new PR that does what @nik9000 original section did, as well as creating a `CoreRequestConverters`.

**Pull request closed without merging**